src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp

```diff
@@ -182,16 +182,41 @@ class Dilithium_PrivateKeyInternal {
       DilithiumPolyVec m_s2;
 };
 
+namespace {
+
+// Return true if randomized signatures were requested
+void validate_dilithium_options(const PK_Signature_Options& options) {
+   BOTAN_ARG_CHECK(!options.using_hash(), "Dilithium does not allow specifying the hash function");
+   BOTAN_ARG_CHECK(!options.using_padding(), "Dilithium does not support padding");
+
+   if(options.using_salt_size()) {
+      BOTAN_ARG_CHECK(!options.using_deterministic_signature(),
+                      "Dilithium cannot support a salt while also being deterministic");
+      BOTAN_ARG_CHECK(options.salt_size().value() == 32, "Dilithium can only be used with a 32 byte salt");
+   }
+
+   // This is available in ML-DSA and might be supported in the future
+   BOTAN_ARG_CHECK(!options.using_prehash(), "Dilithium does not support prehashing");
+}
```

Comment:
Not actually specific to dilithium: I find those explicit parameter checks are a substantial maintenance burden. We'd have to update essentially all algorithms when introducing a new option. And failing to do so, might break users' security assumptions when designing their protocols.

For instance: Let's say we introduce the "context" notion for asym encryptino, _but mess up the sanity check in RSA (or EME-PKCS1)_:

```C++
PK_Encryptor enc(rsa_public_key,
                 PK_Encryptor_Options()
                   .with_padding("PKCS1") // in contrast to OAEP, this won't actually consume the context
                   .with_context("The security of my protocol relies on this string "
                                 "being bound to the ciphertext"));
auto ct = enc.encrypt("my secret message");
```

... in [my aggregate-based example](https://github.com/randombit/botan/pull/4318#issuecomment-2295240036) I had proposed to let algorithms "take" the options they need, rather then "get" them. Read: reset the option after it being read by the consumer.

As a result, the `PK_Options` object can verify that all (or at least all "critical") options have been consumed. Note that we can't do that in `~PK_Options()` as this shouldn't throw. But the constructor of `PK_Signer` could explicitly invoke the verification mechanism after calling `_create_sig_op` on the private key.

```C++
PK_Signer::PK_Signer(Private_Key& key, PK_Options& opts) :
  m_op(key._create_sig_op(opts)) {
  opts.validate_options(key.algo_name()); // -> throws, e.g. "Dilithium did not consume critical option 'context'"
}
```

... or, [if we go full-on builder-pattern](https://github.com/randombit/botan/pull/4318#issuecomment-2297990007):

```C++
PK_Signer PK_Options::build() {
  auto key = this->take_key();
  PK_Signer signer(key, *this);
  this->validate_options(key.algo_name());
  return signer;
}
```

Reply:
With [the extensions I added in the last few days](https://github.com/randombit/botan/pull/4318#issuecomment-2344128175), this has improved much, now.